### ISSUE TITLE
SRT + RTMP playback performance over WebRTC

### ISSUE THREAD

**Issue by llspalex** [AirenSoft/OvenMediaEngine#399]
I have been testing both RTMP and SRT streaming from OBS. With Profile set to 'baseline' or 'main' or 'high' and 'Tune' set to 'zerolatency', the stream plays fine over WebRTC. But if I set the 'Tune' to anything other than 'zerolatency' I get very choppy playback. Presumably this is due to not setting the encoder to bframes=0. 

I'm not sure, but I don't think you can set something like a [Teradek Vidiu GO](https://teradek.com/products/vidiu-go-1?variant=13883839873069#vidiu-pro) (RTMP) or [Videon EdgeCaster](https://videonlabs.com/edgecaster/) to bframes=0. 

My question is, has anyone tested hardware RTMP or SRT encoders with OME playback over WebRTC? And if so, what was your experience? 

Some time ago I tested a Teradek Vidiu GO (which is no longer available to me) streaming RTMP to a Red5Pro server with WebRTC playback, and playback was smooth with the standard Teradek settings, which if I recall correctly don't allow you to set Tune or custom settings like bframes=0.

**Comment by getroot** [comment]
Not supporting H.264's b-frame is not a limitation of OvenMediaEngine, but a limitation of WebRTC browser. Since the default setting of OvenMediaEngine is a bypass setting that passes the input stream as it is, it is recommended to set b-frames=0 in the encoder.

If you want to support an encoder that cannot set bframes=0, use a transcoder in OME. We recommend VP8 encoding for WebRTC.

https://airensoft.gitbook.io/ovenmediaengine/transcoding

**Comment by llspalex** [comment]
Thanks @getroot. Regarding transcoding, is it possible for OME to detect if an incoming stream does not have bframes=0, and only transcode in that instance? We expect the vast majority of incoming streams to be sent via OBS, so they will have bframes=0 already, but it would be nice to support 'non bframes=0' encoders without using blanket transcoding in OME.

Also, if we do enable a VP8 encoder in OME, is it possible to set the stream width, height, bitrate and framerate to 'passthrough' so that the transcoded stream matches the ingest stream as much as possible?

**Comment by getroot** [comment]
It is not possible to dynamically enable the transcoder by detecting b-frames. If you do not have control over the encoder, or if you want to support all encoders regardless of the codec option, you must use the transcoding function.

Let me tell you my experience. There is a commercial media service that we participated in development a while ago, and it was a service that anyone could broadcast. I explained to the streamer as many options as possible, but it was out of control. Sometimes the streamer was sending the video at 30Mbps, the viewers couldn't watch it properly due to their network bandwidth limitations and it looked like a system failure. And we had to pay a lot for network traffic. Therefore, if you cannot control the encoder, you must control the bitrates, resolution, and fps as well as the b-frame option using the transcoder.

And it is possible to get width and height values ​​from the original stream for transcoding options. But bitrates and fps are not easy. Because these are variable values.
